### ISSUE TITLE
Dangerous bug with virtual category module

### ISSUE THREAD

**Issue by Nuranto** [Smile-SA/elasticsuite#2628]
### Preconditions

Magento Version : 2.4.4

ElasticSuite Version :2.10.10

Environment : 
PHP 8.1

### Steps to reproduce

1. Enable virtual categories
2. Load a category page on frontend, for examples : domain.com/iphones
3. try to reach :  domain.com/something/iphones, domain.com/any/thing/really/iphones

### Expected result

1. 404 on both URLs

### Actual result

1. Category iphones gets displayed for both URLs 

Which can be dangerous for SEO. In our case we do have some wrong URLs indexed already (we don't know yet how many and why they get indexed (probably wrong relative links in some cms page)).

If I disable virtualCategory module, problem is solved. I was able to reproduce the error on multiple projects.

**Comment by Nuranto** [comment]
I don't understand the logic of `Smile\ElasticsuiteVirtualCategory\Controller\Router` : Why are we rerouting non-virtual products and categories here ?

But if there is a reason, then I can see that `getCategoryRewrite` is wrong. We should use the whole identifier instead of last chunk because, as it is now : 
1/ We can have wrong URLs like `domain.com/any/thing/really/iphones` that gives a result when it should not.
2/ I bet there is others bug for categories using same url_key like `domain.com/bags/blue` & `domain.com/tshirts/blue`

**Comment by RyseSlade** [comment]
I also have a problem with the latest changes in the `Smile\ElasticsuiteVirtualCategory\Model\Url` class which added the method `getCategoryRewrite`. When the shop is configured to use ".html" suffix for categories the method will load the category and redirect if a rewrite exists. This means now categories can also be called like this https://www.example.com/category but it should only work with https://www.example.com/category.html. Just another SEO issue.

For now I will create a plugin afterGetCategoryRewrite to fix this...

**Comment by romainruaud** [comment]
@RyseSlade instead of creating a plugin in your own project to fix this, can you propose a PR for us here ?

I'll gladly merge it :)

Regards

**Comment by RyseSlade** [comment]
Well, I did not "fix" the issue. I just removed the category rewrite logic with this plugin as I don't need/want that rewrite to happen.

```
public function afterGetCategoryRewrite(Url $subject, $result, $categoryPath, $storeId)
{
    return null;
}
```

**Comment by Nuranto** [comment]
Thanks for the explanation @romainruaud, I now understand the context !

Looking forward for the fix !

**Comment by igorwulff** [comment]
We've decided to just remove the following logic with a patch from the Virtual Category router (vendor/smile/elasticsuite/src/module-elasticsuite-virtual-category/Controller/Router.php):

```
        $categoryRewrite = $this->getCategoryRewrite($identifier);
        if ($categoryRewrite) {
            $request->setAlias(UrlInterface::REWRITE_REQUEST_PATH_ALIAS, $identifier);
            $request->setPathInfo('/' . $categoryRewrite->getTargetPath());

            return $this->actionFactory->create('Magento\Framework\App\Action\Forward');
        }
```
This way we will get an error during a composer install whenever changes have been made in this logic, since this is inherently a bug which I believe we should not fix with plugins or observers as done by @RyseSlade.

**Comment by romainruaud** [comment]
The PR #2743 contains a fix for this one.

I just added a guard fonction early on the Router, to check if we are "under the generated subtree of a virtual category" (see my previous comment for explanations). If that's not the case, we just return null and let the other Magento routers try to handle the request.

Let me know if that's okay for you, this thing should make it fast to our 2.10.12 version.

Regards

**Comment by misthero** [comment]
> @misthero are you using the "Use Categories Path for Product URLs" option ?

nope, in our case 'Use Categories Path for Product URLs' is not checked:

![immagine](https://user-images.githubusercontent.com/3662610/196619639-6ed3d924-0187-41d4-8c18-05e9ea1495ad.png)

reverting to version 2.10.11 from 2.10.12 fixes the problem

**Comment by romainruaud** [comment]
@misthero can you check that the fix in this PR solves your issue ? https://github.com/Smile-SA/elasticsuite/pull/2748

If yes, I'll tag a 2.10.12.1 quickly.

Regards

**Comment by romainruaud** [comment]
@nghiadxvn can you check my PR #2748 ?

**Comment by nghiadxvn** [comment]
@romainruaud yes. It works

**Comment by nghiadxvn** [comment]
Hi @romainruaud, do you have a schedule for releasing this fix?

**Comment by misthero** [comment]
> @misthero can you check that the fix in this PR solves your issue ? #2748
> 
> If yes, I'll tag a 2.10.12.1 quickly.
> 
> Regards

It works!

**Comment by romainruaud** [comment]
2.10.12.1 has been tagged :)

You can update to this version.

I close the issue.
